plan/push_limit.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+func insertLimit(p PhysicalPlan, l *Limit) *Limit {
+	l.SetSchema(p.GetSchema())
+	l.SetChildren(p)
+	p.SetParents(l)
+	return l
+}
+
+// PushLimit implements PhysicalPlan PushLimit interface.
+func (p *Limit) PushLimit(l *Limit) PhysicalPlan {
+	child := p.GetChildByIndex(0).(PhysicalPlan)
+	newChild := child.PushLimit(p)
+	if l != nil {
+		p.Count = l.Count
+		p.Offset = l.Offset
+		p.SetChildren(newChild)
+		newChild.SetParents(p)
+		return p
+	}
+	return newChild
+}
+
+// PushLimit implements PhysicalPlan PushLimit interface.
+func (p *NewSort) PushLimit(l *Limit) PhysicalPlan {
+	child := p.GetChildByIndex(0).(PhysicalPlan)
+	newChild := child.PushLimit(nil)
+	p.ExecLimit = l
+	p.SetChildren(newChild)
+	newChild.SetParents(p)
+	return p
+}
+
+// PushLimit implements PhysicalPlan PushLimit interface.
+func (p *Selection) PushLimit(l *Limit) PhysicalPlan {
+	child := p.GetChildByIndex(0).(PhysicalPlan)
+	newChild := child.PushLimit(nil)
+	if l != nil {
+		return insertLimit(p, l)
+	}
+	p.SetChildren(newChild)
+	newChild.SetParents(l)
+	return p
+}
+
+// PushLimit implements PhysicalPlan PushLimit interface.
+func (p *PhysicalHashSemiJoin) PushLimit(l *Limit) PhysicalPlan {
+	lChild := p.GetChildByIndex(0).(PhysicalPlan)
+	rChild := p.GetChildByIndex(1).(PhysicalPlan)
+	var newLChild, newRChild PhysicalPlan
+	if p.WithAux {
+		newLChild = lChild.PushLimit(l)
+		newRChild = rChild.PushLimit(nil)
```

Comment:
push nil continue the push limit process. Maybe it will bring another limit in future.